Feature/document resources

boralyl opened the pull request:
This will create a section named 'Resource Objects' at the end of the
api documentation that documents all objects used in requests/responses.

It will also add an inline reference link to the documentation next to
the parameter in the api documentation.

_**Example of the auto added inline reference:**_

![screen shot 2018-05-15 at 4 10 13 pm](https://user-images.githubusercontent.com/109865/40088144-90eef00c-585a-11e8-96db-17c40efcdd95.png)

_**Example of the linked reference that documents the object:**_

![screen shot 2018-05-15 at 4 18 46 pm](https://user-images.githubusercontent.com/109865/40088391-c5b63074-585b-11e8-8611-09fcb34c2772.png)

Commit:
Adding the ability to document object's as resources.

This will create a section named 'Resource Objects' at the end of the
api documentation that documents all objects used in requests/responses.

It will also add an inline reference link to the documentation next to
the parameter in the api documentation.

Commit:
Removing get_class_name function.

boralyl commented on doctor/docs/base.py:
```diff
@@ -251,6 +274,51 @@ def get_json_lines(annotation: ResourceAnnotation, field: str, route: str,
+
+    :returns: A list of lines required to generate the documentation.
+    """
+    lines = ['Resource Objects', '----------------']
```
Thought's on this name for the heading?  I don't hate it, but don't love it either.

noonat replied:
I'm also ambivalent about it, but I also can't think of anything better so I think we can just stick with this.

noonat commented on doctor/docs/base.py:
```diff
@@ -295,6 +363,19 @@ def prefix_lines(lines, prefix):
+    :param class_name: The name to convert.
+    :returns: The resource name.
+    """
+    s = re.sub('(.)([A-Z][a-z]+)', r'\1 \2', class_name)
```
What case are you expecting this substitution to handle? It looks like the substitution below would handle things, and when I ran the examples from your test case through that, all of them were converted correctly.

boralyl replied:
I forgot to add a test for this case.  This ensures if the class names starts with an acronym in all caps that it splits properly.

e.g. 

```python
>>> class_name_to_resource_name('HTTPResponse')
'HTTP Response'
```

Without the first regex it wouldn't alter the name at all and would just be `HTTPResponse`.

Commit:
Added another test case for class names that begin with acronyms.

Commit:
Removing the code for documenting object properties.

The reference link now handles this.  This ends up cleaning up the
actual request/response documentation.  The documenting of object
properties was also problematic if an object was in a query parameter or
if a property was also an object.

This also means we can remove our dependency on our forked version of
httpdomain.py

Commit:
Make sure to document properties that are objects too.

Commit:
Don't document resources if there are none to document.

Commit:
Adding tests for no resources to document.
